fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggThroughJoinOnPkFk.java

```diff
@@ -0,0 +1,317 @@
+        if (newGroupBySlots == null || newOutput == null) {
+            return null;
+        }
+        return agg.withGroupByAndOutput(ImmutableList.copyOf(newGroupBySlots), ImmutableList.copyOf(newOutput));
+    }
+
+    private @Nullable Set<Expression> constructNewGroupBy(LogicalAggregate<?> agg, Set<Slot> primaryOutputs,
+            Map<Slot, Slot> primaryToForeignBiDeps) {
+        Set<Expression> newGroupBySlots = new HashSet<>();
+        for (Expression expression : agg.getGroupByExpressions()) {
+            if (!(expression instanceof Slot)) {
+                return null;
+            }
+            if (primaryOutputs.contains((Slot) expression)
+                    && !primaryToForeignBiDeps.containsKey((Slot) expression)) {
+                return null;
+            }
+            expression = primaryToForeignBiDeps.getOrDefault(expression, (Slot) expression);
+            newGroupBySlots.add(expression);
+        }
+        return newGroupBySlots;
+    }
+
+    private @Nullable List<NamedExpression> constructNewOutput(LogicalAggregate<?> agg, Set<Slot> primaryOutput,
+            Map<Slot, Slot> primaryToForeignDeps, FuncDeps funcDeps) {
+        List<NamedExpression> newOutput = new ArrayList<>();
+        for (NamedExpression expression : agg.getOutputExpressions()) {
+            // There are three cases for output expressions:
+            // 1. Slot: the slot is from primary plan, we need to replace it with
+            //             the corresponding slot from foreign plan,
+            //             or skip it when it isn't in group by.
+            // 2. Count: the count is from primary plan,
+            //             we need to replace the slot in the count with the corresponding slot
+            //             from foreign plan
+            // 3. Others: the expression is not from primary plan, we need to keep it
+            if (expression instanceof Slot && primaryToForeignDeps.containsKey(expression)) {
+                expression = primaryToForeignDeps.getOrDefault(expression, expression.toSlot());
+            }
+            if (expression instanceof Alias
+                    && expression.child(0) instanceof Count
+                    && expression.child(0).child(0) instanceof Slot) {
+                // count(slot) can be rewritten by circle deps
+                Slot slot = (Slot) expression.child(0).child(0);
+                if (primaryToForeignDeps.containsKey(slot)
+                        && funcDeps.isCircleDeps(
+                                ImmutableSet.of(slot), ImmutableSet.of(primaryToForeignDeps.get(slot)))) {
+                    expression = (NamedExpression) expression.rewriteUp(e ->
+                            e instanceof Slot
+                                    ? primaryToForeignDeps.getOrDefault((Slot) e, (Slot) e)
+                                    : e);
+                }
+            }
+            if (!(expression instanceof Slot)
```

Comment:
!(expression instanceof Slot) is not necessary? if we support non-standard group by in future, slot may not appear in group by key. refer to: https://dev.mysql.com/doc/refman/8.4/en/sql-mode.html#sqlmode_only_full_group_by

Reply:
I'm not sure if it is guaranteed in this version.

Reply:
done